### DISCUSSION TITLE
Does ZRAM or ZSWAP makes sense in WSL2?

### DISCUSSION THREAD

**Discussion by crramirez** [microsoft/WSL#10862] [Ideas] [answered]
Windows has its own RAM compression system, and WSL2 ultimately runs inside Windows.

Does the Windows RAM manager compresses the Hyper-V RAM as well?

**Comment by crramirez** [comment]
Anyone?

**Comment by 4-FLOSS-Free-Libre-Open-Source-Software** [comment, marked as answer]
ZRAM works very well inside WSL2. In the meantime, you could use the zRAM enabled wsl2 kernel https://github.com/Locietta/xanmod-kernel-WSL2?tab=readme-ov-file#usage

**Reply by crramirez** [reply]
Thank you for sharing. I'll test it. As I couldn't find an answer on the internet if the Hyper-V memory is compressed by the host, today after your answer I asked to ChatGPT and the following is the answer that I received:

No, the memory used by a Hyper-V virtual machine (VM) is generally not subject to compression by the host’s memory compression feature in Windows 10 and Windows 11. Here’s why:
1. Hyper-V’s Memory Allocation: Hyper-V operates at a lower level than typical user-mode applications. It uses a Type-1 hypervisor architecture, meaning the hypervisor runs directly on the hardware beneath the host operating system. The memory allocated to VMs is reserved and managed by the hypervisor, not the host OS’s memory manager.
2. Isolation from Host Memory Management: The memory used by VMs is isolated from the host’s standard memory management mechanisms. This isolation ensures that the VMs have consistent and predictable performance, which is crucial for running guest operating systems and applications reliably.
3. Memory Compression Scope: Windows memory compression is designed to work with the memory pages managed by the host’s operating system processes. It compresses the memory of user-mode applications to optimize RAM usage when the system is under memory pressure. Since the VM’s memory is managed by Hyper-V and not by the host OS’s typical memory manager, it doesn’t participate in this compression.
4. Potential Performance Impact: Compressing VM memory at the host level could introduce latency and unpredictability, negatively affecting the performance of the VMs. Hyper-V aims to provide VMs with as close to raw hardware performance as possible, so adding host-level compression would counteract this goal.
5. Dynamic Memory Feature: Hyper-V offers its own memory management feature called Dynamic Memory, which allows Hyper-V to adjust the amount of memory available to a VM based on its needs. This feature operates independently of the host’s memory compression and is designed specifically for virtualized environments.

References and Further Reading:
	•	[Microsoft Docs - Hyper-V Dynamic Memory Overview](https://docs.microsoft.com/en-us/windows-server/virtualization/hyper-v/manage/manage-hyper-v-dynamic-memory)
	•	[Understanding Memory Compression in Windows](https://techcommunity.microsoft.com/t5/windows-blog-archive/windows-10-memory-compression/ba-p/425008)

In summary, the host’s memory compression feature does not compress the memory used by Hyper-V virtual machines. The VM memory is managed separately by Hyper-V to ensure optimal performance and isolation.

**Reply by ksnishat** [reply]
![image](https://github.com/user-attachments/assets/33587e26-ad1b-4c11-84e8-8e430f8390b5)

![1](https://github.com/user-attachments/assets/f44a5ab0-cfa0-4726-aa3f-fce4a9d747be)
I could not set zram after installing this custom kernel. Can you please provide a small set of instructions?

**Reply by 4-FLOSS-Free-Libre-Open-Source-Software** [reply]
Zram confirmed working

**Reply by mangkoran** [reply]
@4-FLOSS-Free-Libre-Open-Source-Software May we know how to enable/add zram in WSL?

**Reply by Artoria2e5** [reply]
A kernel that has zram or zswap included is not all it takes to get it! You gotta explicitly enable it.

For how to get zram going, https://wiki.archlinux.org/title/Zram should be it. It's a bunch of commands to set up though.

Aaaand that's why I recommend just using zswap. Add `kernelCommandLine="zswap.enabled=1 zswap.compressor=zstd zswap.zpool=zsmalloc"` to your wslcondig under [wsl2] and it should work. I'm going to test this tomorrow because Haskell Language Server keeps crashing the VM by blowing up the RAM.

**Reply by 4-FLOSS-Free-Libre-Open-Source-Software** [reply]
> ![image](https://private-user-images.githubusercontent.com/68187372/397416020-33587e26-ad1b-4c11-84e8-8e430f8390b5.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjI3NzQ1NDMsIm5iZiI6MTc2Mjc3NDI0MywicGF0aCI6Ii82ODE4NzM3Mi8zOTc0MTYwMjAtMzM1ODdlMjYtYWQxYi00YzExLTg0ZTgtOGU0MzBmODM5MGI1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTEwVDExMzA0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIzNWI1NzE2NTJkY2ZjM2NlNmVlZWRiZDRiMzU1NDM4Mjg1MTFiZmI2ZTVhY2UzMWU2MDRmNmM3MzY5NWI1NmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.OdaQ__T0IZz1Suzl8gpCZOR7Sluyc6SSv0taGdm-8NQ)
> 
> ![1](https://private-user-images.githubusercontent.com/68187372/397415671-f44a5ab0-cfa0-4726-aa3f-fce4a9d747be.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjI3NzQ1NDMsIm5iZiI6MTc2Mjc3NDI0MywicGF0aCI6Ii82ODE4NzM3Mi8zOTc0MTU2NzEtZjQ0YTVhYjAtY2ZhMC00NzI2LWFhM2YtZmNlNGE5ZDc0N2JlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTEwVDExMzA0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc3MGJkNDY0MWIyOTJmYTY2MTlhNzBkOGVkODI5MTk1YmFiYzViNGM4NGQwOGNmNGM0YjJiYmFhNDhiMmUyNzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.8-N6BCRR6JUVwfx3e7_hpZmVogIfD5GyJxMGP9dWf5A) I could not set zram after installing this custom kernel. Can you please provide a small set of instructions?

The configuration files is the wrong place. 
No idea what you looking for. 
No ZRAM options exists there. 




> @4-FLOSS-Free-Libre-Open-Source-Software May we know how to enable/add zram in WSL?

https://github.com/Locietta/xanmod-kernel-WSL2/issues/142#issuecomment-3511030501




> Aaaand that's why I recommend just using zswap. Add `kernelCommandLine="zswap.enabled=1 zswap.compressor=zstd zswap.zpool=zsmalloc"` to your wslcondig under [wsl2] and it should work. I'm going to test this tomorrow 

Yes, zwap works in xanmod WSL2 Kernel as described. Just add in .wslconf `kernelCommandLine="zswap.enabled=1" `

**Comment by Artoria2e5** [comment]
@mangkoran Alright as promised, guide to getting zswap turned on (zram involves per-distro configuration (need to set up startup tasks!) and might cause inversion BS with WSL2's disk swap file)

1. get xanmod-kernel-WSL2 using the scoop command
2. `code ~/.wslconfig` to add
   ```ini
   [wsl2]
   kernelCommandLine="zswap.enabled=1 zswap.shrinker_enabled=1 zswap.compressor=zstd zswap.zpool=zsmalloc vm.swappiness=133"
   [experimental]
   autoMemoryReclaim=DropCache
   ```
3. `wsl --shutdown`
4. open your favorite wsl instance, I don't care
5. Run `grep -r . /sys/module/zswap/parameters/` and check that you have `/sys/module/zswap/parameters/enabled:Y`
6. Do something that eats up your memory, e.g. open haskell language server
7. Run `grep -r . /sys/kernel/debug/zswap/` to see how much zswap is saving you

Also useful links are:
* https://wiki.archlinux.org/title/Zswap, for general linux stuff about zswap
* https://www.whitewaterfoundry.com/blog/2025/3/7/get-more-from-less-how-zswap-optimizes-memory-in-wsl2, which is where the gnarly-looking vm.swappiness=133 came from
  * there's also a nice script

### Why high swappiness?
Swappiness determines whether the kernel prioritizes emptying the file pages (by writing to disk, usually) or emptying the anonymous pages (by moving them to swap) when memory is short. As a result the ideal swappiness is (100 &times; (IO cost of emptying file page) / (IO cost of emptying anonymous page)).

The IO cost of emptying a file page comes two things:
* Potentially needing to write the disk if it represents a pending write, one write per page-emptying max.
* Potentially needing to read the disk once again if the file is needed later, one read per page-emptying cycle max.

The IO cost of emptying an anonymous page also comes from two things:
* Definitely needing to write the page to disk.
* Potentially needing to read the page from disk if it's needed later, one read per page-emptying cycle max.

The kernel default swappiness is 60, which is very reasonable assuming the same IO device is used for file and swap. After all it's two potential costs against one definitive cost plus one potential cost.

However, zswap and zram lowers the *effective* cost of reading and writing to swap. Zswap does it by acting as a LRU cache in front of swap. Zram does it by acting *as* the swap. So our cost for emptying an anonymous page becomes:

* Needing to store the page somewhere, which can be an area of RAM or the disk
* Potentially needing to read the page from an area of RAM or the disk

This creates a situation where triggering the swap system may be cheaper than file IO even without considering filesystem overheads. It is totally possible for a page to be instead put somewhere in the RAM, which for most intents and purposes have near zero IO cost compared to disk.

On top of that we have filesystem and other overheads for file IO. In the case of WSL, well we all know about the overhead of accessing Windows drives via "drvfs". So what does decreased swap IO cost and increased file IO cost mean? That's right, higher swappiness is only logical.

### Example performance values
Workload: VS Code and Haskell Language Server

```sh-session
# ./zswap_stats
Zswap Statistics:
-----------------
Stored Pages: 502402
Pool Total Size: 533MB
Uncompressed RAM: 2.0GB
RAM Saved: 1.4GB
# free -h
               total        used        free      shared  buff/cache   available
Mem:           3.8Gi       3.6Gi       171Mi       1.1Mi       271Mi       226Mi
Swap:          4.0Gi       2.6Gi       1.4Gi
# grep -r . /sys/kernel/debug/zswap/
/sys/kernel/debug/zswap/stored_pages:502170
/sys/kernel/debug/zswap/pool_total_size:558354432
/sys/kernel/debug/zswap/written_back_pages:154615
/sys/kernel/debug/zswap/reject_compress_poor:0
/sys/kernel/debug/zswap/reject_compress_fail:505
/sys/kernel/debug/zswap/reject_kmemcache_fail:0
/sys/kernel/debug/zswap/reject_alloc_fail:0
/sys/kernel/debug/zswap/reject_reclaim_fail:143
/sys/kernel/debug/zswap/pool_limit_hit:0
# grep -r . /sys/module/zswap/parameters/
/sys/module/zswap/parameters/enabled:Y
/sys/module/zswap/parameters/shrinker_enabled:Y
/sys/module/zswap/parameters/max_pool_percent:20
/sys/module/zswap/parameters/compressor:zstd
/sys/module/zswap/parameters/zpool:zsmalloc
```

**Reply by 4-FLOSS-Free-Libre-Open-Source-Software** [reply]
Not having zswap.compressor=zstd currently only lz4 supported.

- https://github.com/Locietta/xanmod-kernel-WSL2/pull/94#discussion_r1864556770

**Reply by Artoria2e5** [reply]
> Not having zswap.compressor=zstd currently only lz4 supported.
> 
>     * [[Kenrel Config] enable more features Locietta/xanmod-kernel-WSL2#94 (comment)](https://github.com/Locietta/xanmod-kernel-WSL2/pull/94#discussion_r1864556770)

No, CONFIG_KERNEL_ZSTD=y is present, which allows zstd to be used when you ask for it.

**Comment by crramirez** [comment]
There is a new version of the script that compiles the official Microsoft Kernel adding zswap. It was updated for the kernel 6.x case with external vhdx modules disk.

https://github.com/dhanar10/wsl2-kernel-zswap
